### DISCUSSION TITLE
Allow `injectScript` to target a subset of pages

### DISCUSSION THREAD

**Discussion by Alexandre-Fernandez** [withastro/roadmap#365] [Proposal]
https://docs.astro.build/en/reference/integrations-reference/#injectscript-option
`injectScript` should have an additional parameter to choose what page to inject the script in, this would allow better optimized integrations.

**Comment by delucis** [comment]
Picking this up and fleshing out the idea a little.

Currently, `injectScript` supports injecting client-side JS (bundled and unbundled) and server-side imports into an Astro project. This happens across all routes/pages.

Sometimes it would be useful to be able to `injectScript` and only target a subset of your routes and pages. For example, you may wish to only inject a client-side auth library in `/admin` routes, use Tailwind’s injected styles only in a marketing area of the site, or add specific analytics scripts to a landing page.

We don’t need to decide on APIs during the early stages of discussion, but could look to how [`injectRoute`](https://docs.astro.build/en/reference/integrations-reference/#injectroute-option) and the [upcoming redirects API](https://github.com/withastro/roadmap/pull/587) work for inspiration.

**Reply by andremralves** [reply]
Hey, I liked the idea. I thought about being able to pass an array of targets to the API, something like:

```js
injectScript('page', 'alert("Hello from Astro")', ['admin', 'index.astro']);
```
or

```js
injectScript({
  stage: 'page',
  content: 'alert("Hello from Astro")', 
  targets: ['admin', 'index.astro']
});
```

But what got me thinking is how should we handle the generated script files? Currently all the Injected Scripts are added to a single file `page.js` and all pages import this file with something like: `<script type="module" src="/_astro/page.8ec61a0e.js"></script>`.

So if we add targets to the scripts, I guess it would be necessary to create at least one `.js` file for each target. For example:

If we have two `injectScripts`, one for all pages and the other just for `/admin`. The pages out of `/admin` would import just `page.js` and the pages within `/admin` would import both `page.js` and the script file specific for `/admin` pages, e.g. `page_admin.js`.

Do you think the same, or there is an alternative?

**Reply by delucis** [reply]
It would likely follow the same bundling logic we already use for scripts on different pages. The reason all injected scripts end up in one file currently is because they are all on all pages. If they were injected selectively, I believe the Vite bundler should be able to split those up sensibly for us.
